### PR TITLE
core/state/snapshot: implement storage iterator

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR implements the storage iterator for snapshot.

Note: it has passed all consensus tests and it's verified in `goerli` testnet

### CODE REVIEW

File: core/state/snapshot/iterator_binary.go [hunk 1]
```diff
@@ -92,30 +148,62 @@ first:
...
+func (it *binaryIterator) Slot() []byte {
+	if it.accountIterator {
+		return nil
+	}
+	blob, err := it.a.(*diffStorageIterator).layer.Storage(it.account, it.k)
+	if err != nil {
+		it.fail = err
+		return nil
+	}
+	return common.CopyBytes(blob)
```
Comment (holiman) [inline core/state/snapshot/iterator_binary.go:188]: Careful, I think this converts `nil` into empty slice, which makes the loop in fast iterator leak out empty values

File: core/state/snapshot/iterator_test.go [hunk 2]
```diff
@@ -134,10 +194,12 @@ func verifyIterator(t *testing.T, expCount int, it AccountIterator) {
 		if bytes.Compare(last[:], hash[:]) >= 0 {
 			t.Errorf("wrong order: %x >= %x", last, hash)
 		}
-		if it.Account() == nil {
+		count++
+		if verify == verifyAccount && it.(AccountIterator).Account() == nil {
+			t.Errorf("iterator returned nil-value for hash %x", hash)
+		} else if verify == verifyStorage && it.(StorageIterator).Slot() == nil {
```
Comment (holiman) [inline core/state/snapshot/iterator_test.go:200]: Might be good to check not only against `nil`, but also count empty slice as nil?

File: core/state/snapshot/iterator.go [hunk 3]
```diff
@@ -203,3 +223,175 @@ func (it *diskAccountIterator) Release() {
...
+// Next steps the iterator forward one element, returning false if exhausted.
+func (it *diskStorageIterator) Next() bool {
+	// If the iterator was already exhausted, don't bother
+	if it.it == nil {
+		return false
+	}
+	// Try to advance the iterator and release it if we reached the end
+	prefix := append(rawdb.SnapshotStoragePrefix, it.account.Bytes()...)
+	for {
+		if !it.it.Next() || !bytes.HasPrefix(it.it.Key(), prefix) {
```
Comment (holiman) [inline core/state/snapshot/iterator.go:356]: Is this check really needed: `!bytes.HasPrefix(it.it.Key(), prefix)` ?
Why would the key not have the correct prefix, assuming that `NewIterator` is correct?

File: core/state/snapshot/iterator_binary.go (hunk 1 above)
Comment (karalabe) [reply core/state/snapshot/iterator_binary.go:188]: It does not.

File: core/state/snapshot/iterator.go [hunk 4]
```diff
@@ -203,3 +223,175 @@ func (it *diskAccountIterator) Release() {
...
+// diffStorageIterator is a storage iterator that steps over the specific storage
+// (both live and deleted) contained within a single diff layer. Higher order
+// iterators will use the deleted slot to skip deeper iterators.
+type diffStorageIterator struct {
+	// curHash is the current hash the iterator is positioned on. The field is
+	// explicitly tracked since the referenced diff layer might go stale after
+	// the iterator was positioned and we don't want to fail accessing the old
+	// hash as long as the iterator is not touched any more.
+	curHash common.Hash
+	account common.Hash
```
[the remark this answers is not in the document]
Comment (rjl493456442) [reply core/state/snapshot/iterator.go:236]: Yes, it's necessary when we try to retrieve slot from snapshot.

**Comment by rjl493456442** [comment]
@karalabe @holiman I updated the PR, please take another look

File: core/state/snapshot/iterator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (karalabe) [reply core/state/snapshot/iterator.go:238]: Ah indeed, I saw it later but forgot to delete this comment.

**Pull request merged**